### PR TITLE
Update mermaid 8.5.0 - Refactor repo

### LINKED ISSUE #39: Update to latest Mermaid release
### Enhancement

This is just a reminder to update the mermaid release used to the latest version when doing a new release. 

Currently mermaid is at release [8.2.4](https://github.com/knsv/mermaid/releases/tag/8.2.4) while we are at [8.0.0](https://github.com/SemanticMediaWiki/Mermaid/pull/24)

### PR DESCRIPTION
**Pull request by tylergibson**
This PR is made in reference to: #39 

This PR addresses or contains:
* Updates to using Mermaid 8.5.0
* Refactors the repo to use new Extension.json schema
* Adds support for all mermaid configuration options
* Enables direct future upstream updates from main Mermaid repo through yarn install

This PR includes:
- [x] Tests (unit/integration)
- [x] CI build passed

Fixes #39

### CODE REVIEW

**Comment by mwjames** [comment]
Without looking at details:

- Moving files from /src to /includes is a no-go, this repo follows PSR-4
- Moving files outside of /res is not acceptable unless of course
there is a technical reason to do so
- Introducing `yarn` as dependency is questionable, required files
have to be included with a release

**Comment by tylergibson** [comment]
Thanks for the feedback @mwjames - I was trying to follow the conventions here - https://www.mediawiki.org/wiki/Manual:Developing_extensions which say to place PHP execution code into /includes. 

Doesn't really matter to me what the dir structure is though :).

Re: Introducing Yarn - The mermaid repo uses Yarn and the mermaid package is an NPM package itself, so you need either npm or yarn (or some other node package manager) to avoid having a static JS dump of Mermaid.js  
This path ensures that, unlike the current release, anyone can reproduce the mermaid.min.js file from origin.

**Comment by mwjames** [comment]
> Thanks for the feedback @mwjames - I was trying to follow the conventions
> here - https://www.mediawiki.org/wiki/Manual:Developing_extensions which say
> to place PHP execution code into /includes.
>
> Doesn't really matter to me what the dir structure is though :).

WMF doesn't follow PSR-4 [0] best practice.

> Re: Introducing Yarn - The mermaid repo uses Yarn and the mermaid package is
> an NPM package itself, so you need either npm or yarn (or some other node
> package manager) to avoid having a static JS dump of Mermaid.js
> This path ensures that, unlike the current release, anyone can reproduce the
> mermaid.min.js file from origin.

As development tool `yarn` could be acceptable (e.g. I don't used and
I don't have node.js available) for the release process but we need
the static files which should reside in a subfolder /res (or
/resources) but not scattered around different top directories.

Any new release you may fetch those filesvia yarn and copy them to the
defined location.

[0] https://www.php-fig.org/psr/psr-4/

File: extension.json [hunk 1]
```diff
@@ -16,10 +17,44 @@
...
+	},
+	"ExtensionMessagesFiles": {
+		"MermaidMagic": "i18n/Mermaid.i18n.php"
+	},
+	"Hooks" : {
+		"ParserFirstCallInit" : "Mermaid\\Hooks::onParserFirstCallInit",
+		"OutputPageParserOutput" : "Mermaid\\Hooks::onOutputPageParserOutput"
+	},
+	"config": {
+		"Theme": "forest"
```
Comment (kghbln) [inline extension.json:57]: No need for breaking change here. Just use "config_prefix" as `"config_prefix": "mermaidg",` and `"DefaultTheme": "forest"` for "config"

**Comment by tylergibson** [comment]
@mwjames I've updated the directory structure to better reflect PSR-4 and PHP-PDP, please have another look.

@kghbln Have made all the changes you noted, have a look and see how they look now

**Comment by mwjames** [comment]
As you might be aware this repo comes with PHPUnit tests which needs
to be passed.

Also, if you require a specif config you add a setter/getter because doing
`$globalConfig = MediaWikiServices::getInstance()->getMainConfig();`
defies the idea of dependency injection (you inject whatever you need
from the top).

Doing `$globalConfig =
MediaWikiServices::getInstance()->getMainConfig();` is a kitchen  sink
that cannot be tested so you have to isolate what you need and inject
that instead of just pulling in unnecessary dependencies.

File: extension.json [hunk 2]
```diff
@@ -1,11 +1,12 @@
...
-	"version": "2.3.0-alpha",
+	"version": "2.4.0-alpha",
 	"author": [
-        	"James Hong Kong"
+			"James Hong Kong",
+			"Tyler Gibson"
 	],
 	"url": "https://github.com/SemanticMediaWiki/Mermaid/",
-	"descriptionmsg": "mermaid-desc",
+	"descriptionmsg": "Mediawiki Extension enabling Mermaid.js diagrams at runtime",
```
[the remark this answers is not in the document]
Comment (tylergibson) [reply extension.json]: Thanks for the feedback, should be all sorted now

**Comment by kghbln** [comment]
> Have made all the changes you noted, have a look and see how they look now

Looks good to me. Thanks!

**Comment by mwjames** [comment]
The way the `elseif` of the config is handled should be isolated as an
extra class to manipulate those available options as parameters.

The new class needs phpunit test.

Overall this should be a 3.0 release given the various changes.

**Comment by kghbln** [comment]
> Overall this should be a 3.0 release given the various changes.

I had the same feeling, but this is release details. I will adjust accordingly then doing the next release after the refactor is fully fluff.

**Comment by kghbln** [comment]
My approval only covers the comments I made which were resolved in the meantime.

**Comment by tylergibson** [comment]
> As you might be aware this repo comes with PHPUnit tests which needs to be passed. Also, if you require a specif config you add a setter/getter because doing `$globalConfig = MediaWikiServices::getInstance()->getMainConfig();` defies the idea of dependency injection (you inject whatever you need from the top). Doing `$globalConfig = MediaWikiServices::getInstance()->getMainConfig();` is a kitchen sink that cannot be tested so you have to isolate what you need and inject that instead of just pulling in unnecessary dependencies.

Ah, I totally forgot about the unit tests.  Sincere apologies.  Also I am extremely rusty with PHP, so any help you may be able to lend with the phpunit setup would be super appreciated.

Re: getMainConfig, this shouldn't be dependency injection at all.  I purposely removed that as there is no need to ever set the mainconfig, only to do a single get-check for the default.  
I am using the MWF defined getter for this, following their prescribed method to do so.
https://www.mediawiki.org/wiki/Manual:Configuration_for_developers
https://www.mediawiki.org/wiki/Manual:Config.php

It looks like we should be using the extension.json to define all of the config features (and to you point, pulling them out of the ifelse insanity I have in there currently)
https://www.mediawiki.org/wiki/Manual:Extension.json/Schema#config

That would get out of the kitchen-sink problem, since it would be a registry for just the scope of the extension, and would provide a known array to iterate over for the values to load into the instance.

I can look at that this weekend unless yall can get to it earlier.

Thanks for all the awesome feedback and help!

**Comment by mwjames** [comment]
> Re: getMainConfig, this shouldn't be dependency injection at all. I purposely removed that as there is no need to ever set the mainconfig, only to do a single get-check for the default.
I am using the MWF defined getter for this, following their prescribed method to do so.
https://www.mediawiki.org/wiki/Manual:Configuration_for_developers
https://www.mediawiki.org/wiki/Manual:Config.php

No, `getMainConfig` shouldn't be part of a class (this where you pull in the dependency).

I just looked at the code, all you do is `$globalConfig->has("mermaidgDefaultTheme")` with the config object hence you need to set the value for `mermaidgDefaultTheme` from the outside not from within the class.

Configs are either isolated (by a dedicated class) or you have a setter but never have the entire $GLOBALS or `Config` object be part of a class.

```
		if ( $globalConfig->has("mermaidgDefaultTheme") ) {
			'ext.mermaid.styles'
			$graphConfig = [
		);
				'theme' => $globalConfig->get("mermaidgDefaultTheme")
```

**Comment by howlowck** [comment]
@mwjames Please take a look at the updates for this PR.

I made the changes based on your requirements (the best that I could 😬):
* Added `ServiceWiring.php` to include two new services: 
  * `Mermaid.Config` (that gets the default theme)
  * `Mermaid.MermaidConfigExtractor` (that abstracts away the `if/elseif` statements)
* Only Invoked `MediaWikiServices::getInstance()` [in `onParseFunction`](https://github.com/SemanticMediaWiki/Mermaid/pull/47/files#diff-067ac20ecdf4315868de966295e0d01bR54-R57) where we are instantiating the class
* Fixed Tests

Disclamer: I'm a PHP dev, but not a MediaWiki dev, so I'm happy to make any more changes that would make the code more idiomatic to the MediaWiki community.

**Comment by howlowck** [comment]
I'm also looking into the failing Travis builds

**Comment by tylergibson** [comment]
@kghbln @mwjames - Pinging for review.  Much thanks to my colleague @howlowck - I believe he has addressed the issues and refactored the repo properly.

File: src/ServiceWiring.php [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+return [
+	/**
+	 * Mermaid.MermaidConfigExtractor
+	 *
+	 * @return callable
+	 */
+	'Mermaid.MermaidConfigExtractor' => function (MediaWikiServices $services) {
+		return new MermaidConfigExtractor();
+	},
```
Comment (JeroenDeDauw) [inline src/ServiceWiring.php:23]: I think using this service wiring stuff is extremely bad practice. Guess you just followed what was recommended on MediaWiki.org though. Lots of stuff to ignore there.

File: tests/phpunit/Unit/MermaidParserFunctionTest.php [hunk 4]
```diff
@@ -13,108 +13,114 @@
...
-			'class="ext-mermaid" data-mermaid="{&quot;content&quot;:&quot;graph LR;&quot;,&quot;config&quot;:{&quot;theme&quot;:&quot;foo&quot;,&quot;flowchart&quot;:{&quot;useMaxWidth&quot;:false}}}"><div class="mermaid-dots"></div></div>'
+			['graph LR;', 'config.theme=foo', 'config.flowchart.useMaxWidth=false'],
+			'<div class="ext-mermaid" data-mermaid="{&quot;content&quot;:&quot;graph LR;&quot;,&quot;config&quot;:{&quot;theme&quot;:&quot;foo&quot;,&quot;flowchart&quot;:{&quot;useMaxWidth&quot;:false}}}"><div class="mermaid-dots"></div></div>'
 		];
 
 	}
 
+	protected function getMockConfig()
+	{
+		$configMock = $parser = $this->createMock('\Mermaid\Config');
```
Comment (JeroenDeDauw) [inline tests/phpunit/Unit/MermaidParserFunctionTest.php:109]: cp error

File: tests/phpunit/Unit/MermaidParserFunctionTest.php [hunk 5]
```diff
@@ -13,108 +13,114 @@
...
+		$configMock->method('getDefaultTheme')->willReturn('forest');
+
+		return $configMock;
+	}
+
+	protected function getMockConfigExtractor()
+	{
+		$valueMap = TestingConsts::EXTRACTOR_VALUE_MAP;
+
+		$extractorMock = $parser = $this->createMock('\Mermaid\MermaidConfigExtractor');
```
Comment (JeroenDeDauw) [inline tests/phpunit/Unit/MermaidParserFunctionTest.php:120]: cp error

File: tests/bootstrap.php [hunk 6]
```diff
@@ -8,11 +8,11 @@
...
 
-if ( !class_exists( 'Mermaid' ) || ( $version = Mermaid::getVersion() ) === null ) {
-	die( "\nMermaid is not available, please check your Composer or LocalSettings (wfLoadExtension).\n" );
-}
+//if ( !class_exists( 'Mermaid' ) || ( $version = Mermaid::getVersion() ) === null ) {
+//	die( "\nMermaid is not available, please check your Composer or LocalSettings (wfLoadExtension).\n" );
+//}
 
-print sprintf( "\n%-20s%s\n", "Mermaid: ", $version );
+//print sprintf( "\n%-20s%s\n", "Mermaid: ", $version );
```
Comment (JeroenDeDauw) [inline tests/bootstrap.php:15]: Why commented out?

File: src/ServiceWiring.php (hunk 3 above)
Comment (JeroenDeDauw) [reply src/ServiceWiring.php:23]: Tools do not understand what is happening here

File: resources/ext.mermaid.js [hunk 7]
```diff
@@ -15,7 +15,7 @@
 
 	var config = mw.config.get( 'mermaid' );
 
-	mw.loader.using( [ 'mediawiki.api', 'ext.mermaid', 'ext.mermaid.theme.' + config.theme ] ).then( function () {
+	mw.loader.using( [ 'mediawiki.api', 'ext.mermaid' ] ).then( function () {
```
Comment (JeroenDeDauw) [inline resources/ext.mermaid.js:19]: What happened with the removed module here?

Comment (JeroenDeDauw) [reply resources/ext.mermaid.js:19]: Ah I see - it was in one of the removed files.

**Comment by JeroenDeDauw** [comment]
Needs a rebase against master due to https://github.com/SemanticMediaWiki/Mermaid/commit/d19485e154580678537648837a6ec6e6c8e06c8f

File: src/ServiceWiring.php (hunk 3 above)
Comment (howlowck) [reply src/ServiceWiring.php:23]: Hey @JeroenDeDauw, thanks for that.. haha yea. I'm new to mediawiki dev, so I just followed the instructions from mediawiki. Looking for guidance on better practice. We want to change the service binding here, or we can make enhancements once the PR is through.

**Comment by howlowck** [comment]
Thanks @JeroenDeDauw! Will rebase today 👍

**Comment by howlowck** [comment]
Rebased upstream

File: src/ServiceWiring.php (hunk 3 above)
Comment (JeroenDeDauw) [reply src/ServiceWiring.php:23]: A follow up PR is welcome but not required.

File: tests/phpunit/Unit/MermaidParserFunctionTest.php (hunk 4 above)
Comment (JeroenDeDauw) [reply tests/phpunit/Unit/MermaidParserFunctionTest.php:109]: Fixed on master

File: tests/phpunit/Unit/MermaidParserFunctionTest.php (hunk 5 above)
Comment (JeroenDeDauw) [reply tests/phpunit/Unit/MermaidParserFunctionTest.php:120]: Fixed on master

**Pull request merged**